sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,161 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"errors"
+	"os"
+	"path"
+
+	"github.com/snapcore/snapd/sandbox/cgroup/inotify"
+)
+
+type appData struct {
+	name        string
+	cgroupPaths []string
+	npaths      int
+	channel     chan string
+}
+
+// This class allows to monitor a CGroup and detect when all the running
+// instances have been closed
+
+type CGroupMonitor struct {
+	watched map[string][]*appData
+	watcher *inotify.Watcher
+	channel chan appData
+}
+
+var currentCGroupMonitor = CGroupMonitor{
+	watcher: nil,
+	channel: make(chan appData),
+	watched: make(map[string][]*appData),
+}
+
+func monitorMainLoop() {
+	for {
+		select {
+		case event := <-currentCGroupMonitor.watcher.Event:
+			if event.Mask&inotify.InDelete != 0 {
```

Comment:
I would vote for moving the bodies of these case blocks into separate functions, as the level of indentation here is already quite high.

Reply:
Done.